Add diags for zfs and zpool properties

ofaaland opened the pull request:
Add diagnostics for check ZFS pool and dataset properties.

All variables are array variables.

Supported zpool variables:
DIAG_ZPOOL_MULTIHOST
DIAG_ZPOOL_AUTOREPLACE

Tests are applied to a pool if the pool name matches the
regex in DIAG_ZPOOL_NAME.

Supported zfs dataset variables:
DIAG_ZFS_RECORDSIZE
DIAG_ZFS_DNODESIZE
DIAG_ZFS_XATTR
DIAG_ZFS_CANMOUNT
DIAG_ZFS_COMPRESSION

Tests are applied to a dataset if the pool name matches the
regex in DIAG_ZFS_DATASET_NAME.

For example:

	DIAG_ZFS_DATASET_NAME[0]=".*/mdt1"
	DIAG_ZFS_DNODESIZE[0]="auto"

	DIAG_ZFS_DATASET_NAME[1]="jet.*/ost1"
	DIAG_ZFS_RECORDSIZE[1]="1M"

will result in the dnodesize being checked on a dataset named
mdt1, in any pool; and in the recordsize being checked on a
dataset named ost1, in a pool whose name starts with "jet".

Fixes #11

ofaaland commented:
@garlick you can ignore this for the moment.  I need to get feedback from Cameron and/or Mike Gilbert first.   These new diagnostics have two different code structures and I want to ask them something about them.  Thanks.

garlick commented:
Will do.  I'm happy to press the button when you are ready but I don't
have a good way to test this right now so hope you and others can do
that and give the thumbs up first.

On Wed, Mar 18, 2020 at 3:22 PM Olaf Faaland <notifications@github.com> wrote:
>
> @garlick you can ignore this for the moment. I need to get feedback from Cameron and/or Mike Gilbert first. These new diagnostics have two different code structures and I want to ask them something about them. Thanks.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

ofaaland commented:
>  I don't have a good way to test this right now so hope you and others can do
> that and give the thumbs up first.

Yes, we'll put it on jet and opal and make sure it's well behaved before I actually ask you to land it.

mkgilbert commented:
@ofaaland , in answer to the question you asked offline, if it was my choice, I would choose the scripting style you used in the `zpool-properties.t` file. I personally find that to be more concise and readable of the two.

cmharr commented:
I personally prefer the structure of the zpool properties, with the multiple functions.

cmharr commented on diags/zfs-properties.t:
```diff
@@ -0,0 +1,117 @@
+			verify_property ${dataset} compression ${DIAG_ZFS_COMPRESSION[$idx]}
+		fi
+	done
+	echo >&2
```
Not sure if you want this echo statement there as it adds a newline under the test count that I don't believe other tests have.

ofaaland replied:
Yes, that's a mistake.  Thank you.

ofaaland replied:
fixed

ofaaland commented:
Thanks @cmharr and @mkgilbert for answering that question, I'll structure both like the zpool-properties.t and resubmit, as well as testing on jet and opal.

cmharr commented on diags/zpool-properties.t:
```diff
@@ -0,0 +1,95 @@
+# number of tests
+#
+num_tests=0
+foreach_pool_and_property count_tests
```
This appears to make the action (arg $4) always be count_tests, which results in no actual comparison happening between the set value and the expected value.

ofaaland replied:
Correct.  With action="count_tests", the tests are not executed.  We just calculate the number of tests.

If you look at line 89, you see that we have to declare, in advance, the number of tests which will be executed.  But the syntax supports different sets of tests, and so the number can only be determined by counting.  For example:
```
DIAG_ZPOOL_NAME[0]=".*/mdt1"
DIAG_ZPOOL_MULTIHOST[0]="on"

DIAG_ZPOOL_NAME[0]=".*/ost1"
DIAG_ZPOOL_MULTIHOST[0]="on"
DIAG_ZPOOL_AUTOREPLACE[0]="on"
```

If run on a jet MDS, with only one pool imported, there would only be 1 test.  But if run on a jet OSS with both pools imported, there would be 4 tests.

So with zpool-properties.t, I handle that with a common function that loops over pools foreach_pool_and_property.  And if action is "count_tests" then the expected/actual values aren't compared, num_tests is incremented.  If action is anything else, then expected/actual values are compared.

ofaaland replied:
Does that change your answer, about which of these two versions you prefer?  This is the key difference.

cmharr replied:
No, I don't think it changes my preference, though perhaps I'm not clear of the scope of this patch. In production, when wanted to verify pool-level attributes, what would you be setting the action to? Just something else?

ofaaland replied:
If you look at line 95, foreach_pool_and_property is called with argument "do_tests".  This results in the actual values being compared with the expected ones.

This file (zpool-properties.t) calls foreach_pool_and_property twice.  First it counts the tests that will be done (line 87) and then declares that number (line 89) and finally actually performs the tests and complains if appropriate (line 95).

cmharr replied:
Oh dear. I had copied and pasted the script to test on zinc and somehow that last line got missed, thus leading to a lot of head scratching. It now makes much more sense.

ofaaland replied:
No problem.  I'm going to re-write zfs-properties to use the same structure, and then put both of them on jet (where it should test stuff) and opal (where it shouldn't break anything).

ofaaland replied:
And in fact I see a bug in zpool-properties.t that I need to fix...

ofaaland replied:
zfs-properties.t is rewritten, and the bugs I found are fixed.

ofaaland commented:
@cmharr I updated zfs-properties.t and fixed the bugs I found in zpool-properties.t.  Take a look if you have a chance - thanks!

ofaaland commented:
Updated because I'd forgotten to implement the "-c" option for spitting out a config based on current values.

ofaaland commented:
Updated to fix a bug in the "-c" option.

I think this is ready for prime time.  Tested successfully on jet.

@cmharr and @garlick please review, thanks.

garlick commented:
Some general questions:

Does `nodediag -c` do something sensible on a system with zfs installed but no dataests or pools?

What happens if you run `nodediag -c` on a system that doesn't have the zfs command installed?  (It looks like it might barf)

The `diagconfig()` function accesses `datasets`, which is a global defined later in the script.  That's a bit odd scoping wise, and I think it works because the whole script is first sourced then the function is called?   Might want to generate the dataset list with a function that is called from both places, e.g. see `ecc.t` and avoid the global.

ofaaland commented:
> Does nodediag -c do something sensible on a system with zfs installed but no dataests or pools?

@cmharr @mkgilbert this seems like a good place for your input.

Currently it emits "1..0 # Skipped:  no ZFS pools" or "... ZFS datasets" on stderr and nothing at all on stdout.   I like that it's not polluting the config and also that it warns the user in case that's not what they expected.  

I could instead go with the "# zfs-properties: no defaults available emitted to stdout", which also seems reasonable to me.

What are your thoughts?

> What happens if you run nodediag -c on a system that doesn't have the zfs command installed? (It looks like it might barf)
Yes, it would have barfed.  Fixed.

> The diagconfig() function accesses datasets, which is a global defined later in the script. That's a bit odd scoping wise, and I think it works because the whole script is first sourced then the function is called? Might want to generate the dataset list with a function that is called from both places, e.g. see ecc.t and avoid the global.
Fixed

cmharr commented:
> Currently it emits "1..0 # Skipped: no ZFS pools" or "... ZFS datasets" on stderr and nothing at all on stdout. I like that it's not polluting the config and also that it warns the user in case that's not what they expected.
> 
> I could instead go with the "# zfs-properties: no defaults available emitted to stdout", which also seems reasonable to me.
> 
> What are your thoughts?

I like that it explicitly says no ZFS pools or datasets found.

ofaaland commented:
@garlick , looks like this is ready for your review when you get a chance.  Testing revealed no problems:
jet
* mgmt node has zfs modules loaded, but no pools - notrun
* mds and oss nodes have modules loaded and pools imported - run, results correct

opal
* no zfs modules loaded - notrun

ofaaland commented:
Yes, please.  Thanks for your help!

garlick commented:
Thanks to you for making it harder to poke your eye out with zfs _in parallel_.  The world is a safer place!
